### PR TITLE
Various fixes to backup solution

### PR DESCRIPTION
**Pull request by NoxHarmonium**
We were having some issues getting [the AWS reference efs-to-efs backup solution](https://docs.aws.amazon.com/solutions/latest/efs-to-efs-backup/welcome.html) to work out of the box in our AWS environment so we made some fixes to get it working that can be shared with the wider community.

I hope these fixes can help others!

**New Features**
- Adds support for HTTP proxies
- Adds support for adding a keypair to the backup EC2 instance
- Can now use three mount points for the backup EFS volume
- Can attach a custom security group to the backup EC2 instance
- Fix issue where backup script would exit before reporting finished state to orchestrator
- Made backup script more robust
- Misc. cleanup and various small fixes

**Still Needs To Be Done**
- I couldn't get the tests to work but I don't have time to debug that at the moment
- I haven't looked at the restore scripts yet, they probably won't work with a proxy. I'll have to look at that later.

### CODE REVIEW

File: source/scripts/efs-backup-fpsync.sh [hunk 1]
```diff
@@ -36,8 +36,8 @@ sudo yum -y install nfs-utils
 
 echo "-- $(date -u +%FT%T) -- sudo yum -y groupinstall 'Development Tools'"
 sudo yum -y groupinstall "Development Tools"
-echo "-- $(date -u +%FT%T) -- wget https://s3.amazonaws.com/%TEMPLATE_BUCKET_NAME%/efs-backup/latest/fpart.zip"
-wget https://s3.amazonaws.com/%TEMPLATE_BUCKET_NAME%/efs-backup/latest/fpart.zip
+echo "-- $(date -u +%FT%T) -- wget https://s3.amazonaws.com/solutions-reference/efs-backup/latest/fpart.zip"
+wget https://s3.amazonaws.com/solutions-reference/efs-backup/latest/fpart.zip
```
Comment (pearj) [inline source/scripts/efs-backup-fpsync.sh]: Probably want to keep `%TEMPLATE_BUCKET_NAME%` here

**Comment by NoxHarmonium** [comment]
I just realised that while I was working on this, a change went through to master which will probably conflict with this PR. I will have to rebase and integrate the latest v1.1 changes.

**Comment by gsingh04** [comment]
thanks for providing the PR, we will look to incorporate these in v1.2

**Comment by NoxHarmonium** [comment]
Thanks for the update @gsingh04! 

I have just rebased my changes on top of master and removed some redundant commits. I'm wondering if I should remove commit 2a0f9ad460a2065d98489013ffb3a30a0a734501 because it might have just been a problem with our environment with all the proxy requirements etc. I will need someone else to verify in their environment.

**Comment by gsingh04** [comment]
@NoxHarmonium verified 2a0f9ad, worked fine for me. Also, proxies should not be a problem when pulling instance metadata information. Did you make sure your instance had a IAM role attached to it when running the script?

**Comment by NoxHarmonium** [comment]
@gsingh04 I triple checked that I had the correct IAM role attached to the instance but I still wasn't able to get it to work. It turned out to be that I did not put the AWS management API IP address into the "no_proxy" environment variable.

This requirement is outlined in the AWS documentation at https://docs.aws.amazon.com/cli/latest/userguide/cli-http-proxy.html#cli-http-proxy-ec2

I have removed that block of code because it should not be needed. I have also fixed up the removed template variables mentioned by @pearj.

**Comment by quiver** [comment]
- Adds support for adding a keypair to the backup EC2 instance
- Can now use three mount points for the backup EFS volume

+1

EFS mount points are really important.

AWS recommends to create mount points for every AZ but CFn template creates only two.

Imagine this scenario.

If you specify more than two subsets as CFn parameters, 
- Auto Scaling Group uses all of them
- EFS mount points use two of them

If instances are launched from non-supported AZ, backup job fails.

It took me a while to debug why EFS-to-EFS solution fails randomly  :-(

**Comment by gsingh04** [comment]
Thanks for your feedback. Please try our latest v1.3 release and let us know the performance improvement or if there are any concerns. 
- Adding key pair is a change that can be done at customer end. We do not recommend adding key pair as we provide Cloudwatch dashboard and logs in S3 to debug issues.
- Some of the regions have only 2 AZs currently, like Mumbai, Canada. EFS is not available in these regions for now, but we did not want the solution to break once EFS is available in these regions; hence the decision to go with 2 AZs

**Comment by shsenior** [comment]
updated with v1.3.1

**Pull request closed without merging**